### PR TITLE
LibGUI+Userland: Make `GUI::ColorPicker` interactive

### PR DESCRIPTION
**Pull request by vkoskiv**
I added a callback to `GUI::ColorPicker` that clients can optionally use to receive real-time color updates, and applied this new feature to every location in the system that uses it, namely:
* Spreadsheet
* PixelPaint
* Magnifier
* Snake
* `LibGUI` itself, in `GUI::ColorInput`, which is used in many more places.

Quick demo:

https://github.com/SerenityOS/serenity/assets/20117975/b65fe275-59e7-4d8c-a609-4bb63172ebc9

### CODE REVIEW

File: Userland/Games/Snake/Game.h [hunk 1]
```diff
@@ -34,6 +34,7 @@ class Game
     Function<bool(u32)> on_score_update;
 
     void set_skin_color(Color);
+    Gfx::Color get_skin_color(void) const { return m_snake_color; }
```
Comment (trflynn89) [inline Userland/Games/Snake/Game.h:37]: Don't need `void` in the param list

File: Userland/Games/Snake/main.cpp [hunk 2]
```diff
@@ -108,10 +108,17 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
...
         if (!was_paused)
             game.pause();
-        auto dialog = GUI::ColorPicker::construct(Gfx::Color::White, window);
+        auto set_skin_color = [&game](Gfx::Color color) {
+            Config::write_u32("Snake"sv, "Snake"sv, "BaseColor"sv, color.value());
+            game.set_skin_color(color);
+        };
+        auto dialog = GUI::ColorPicker::construct(game.get_skin_color(), window);
+        dialog->on_color_changed = [&set_skin_color](Gfx::Color color) {
+            dbgln("Set color to {}", color);
```
Comment (trflynn89) [inline Userland/Games/Snake/main.cpp:117]: Leftover debug?

File: Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp [hunk 3]
```diff
@@ -495,20 +495,70 @@ void SpreadsheetWidget::change_cell_static_color_format(Spreadsheet::FormatType
 {
     VERIFY(current_worksheet_if_available());
 
-    auto dialog = GUI::ColorPicker::construct(Color::White, window(), "Select Color");
-    if (dialog->exec() == GUI::Dialog::ExecResult::OK) {
+    auto preview_color_in_selected_cells = [this, format_type](Gfx::Color color) {
```
Comment (trflynn89) [inline Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp:498]: I'll let @alimpfard review this file 😅

**Review by trflynn89 — COMMENTED** [review]
Very cool!

**Review by LucasChollet — COMMENTED** [review]
I really like the change, but it lacks a bit of consistency on the caller side IMO.
If I read you correctly, the `ExecResult::OK` branch shouldn't be used to render a new color, but only to save the final state (like writing to `Config`).

File: Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp (hunk 3 above)
Comment (vkoskiv) [reply Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp:498]: Yeah... It turned into a bit of lambda spaghetti :sweat_smile:

File: Userland/Games/Snake/main.cpp [hunk 4]
```diff
@@ -108,10 +108,17 @@ ErrorOr<int> serenity_main(Main::Arguments arguments)
         auto was_paused = game.is_paused();
         if (!was_paused)
             game.pause();
-        auto dialog = GUI::ColorPicker::construct(Gfx::Color::White, window);
+        auto set_skin_color = [&game](Gfx::Color color) {
+            Config::write_u32("Snake"sv, "Snake"sv, "BaseColor"sv, color.value());
```
[the remark this answers is not in the document]
Comment (vkoskiv) [reply Userland/Games/Snake/main.cpp:112]: Agreed, will fix, *but* in my defense, this exact case was accounted for in my PR a while back: #10473 :^)

**Comment by vkoskiv** [comment]
> I really like the change, but it lacks a bit of consistency on the caller side IMO. If I read you correctly, the `ExecResult::OK` branch shouldn't be used to render a new color, but only to save the final state (like writing to `Config`).

Agreed. I'll set both the `ExecResult` branches to call `on_color_changed`. The `ExecResult::Cancel` will call it with the originally provided color, thus restoring it to what it was before, and the `ExecResult::OK` branch will call it with the currently selected color.
This way, clients can choose to only care about one of these APIs. There's still a bit of awkward overlap, but I think your suggestion of checking for `ExecResult::OK` to write to config makes the most sense.

File: Userland/Games/Snake/main.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (LucasChollet) [reply Userland/Games/Snake/main.cpp:112]: There is still an issue, imagine you open the color picker to test some colors and then your app crash. Next start-up, you have a weird color you never asked for.

**Comment by vkoskiv** [comment]
All fixed up now. Thanks for the feedback @LucasChollet! This ended up looking far better with these fixes :sunglasses:

File: Userland/Applications/PixelPaint/PaletteWidget.cpp [hunk 5]
```diff
@@ -32,14 +32,18 @@ class ColorWidget : public GUI::Frame {
     virtual void mousedown_event(GUI::MouseEvent& event) override
     {
         if (event.modifiers() & KeyModifier::Mod_Ctrl) {
-            auto dialog = GUI::ColorPicker::construct(m_color, window());
-            if (dialog->exec() == GUI::Dialog::ExecResult::OK) {
-                m_color = dialog->color();
+            auto update_widget = [this](Gfx::Color color) {
```
Comment (LucasChollet) [inline Userland/Applications/PixelPaint/PaletteWidget.cpp:35]: No need to have two lambdas here

Comment (vkoskiv) [reply Userland/Applications/PixelPaint/PaletteWidget.cpp]: Fixed!

File: Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp [hunk 6]
```diff
@@ -495,21 +495,71 @@ void SpreadsheetWidget::change_cell_static_color_format(Spreadsheet::FormatType
...
+        for (auto& position : current_worksheet_if_available()->selected_cells()) {
+            auto* cell = current_worksheet_if_available()->at(position);
+            auto previous_type_metadata = cell->type_metadata();
+            if (format_type == Spreadsheet::FormatType::Background)
+                cell->type_metadata().static_format.background_color = color;
+            else
+                cell->type_metadata().static_format.foreground_color = color;
+            update();
+        }
+        window()->set_modified(true);
```
Comment (MacDue) [inline Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp:508]: You probably don't want to set the window as in the preview:
```suggestion
```

Comment (vkoskiv) [reply Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp]: Fixed!

File: Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp (hunk 3 above)
Comment (alimpfard) [reply Userland/Applications/Spreadsheet/SpreadsheetWidget.cpp:498]: Lambda soup, but sure, LGTM.

**Pull request merged**